src/Microsoft.VisualStudio.ProjectSystem.Managed/Tree/Dependencies/Snapshot/TargetedDependenciesSnapshot.cs

```diff
@@ -174,35 +174,160 @@ void Add(AddDependencyContext context, IDependencyModel dependencyModel)
             Catalogs = catalogs;
             DependenciesWorld = dependenciesWorld;
 
-            bool hasVisibleUnresolvedDependency = false;
-            ImmutableArray<IDependency>.Builder topLevelDependencies = ImmutableArray.CreateBuilder<IDependency>();
+            bool hasVisibleUnresolvedDependency;
 
-            foreach ((string id, IDependency dependency) in dependenciesWorld)
+            // Perform a single pass through dependencies, gathering as much information as possible
+            (TopLevelDependencies, _topLevelDependencyByPath, hasVisibleUnresolvedDependency) = Scan(dependenciesWorld, targetFramework);
+
+            if (hasVisibleUnresolvedDependency)
+            {
+                // Walk the dependency graph to find visible, unresolved dependencies which are reachable
+                // from visible top-level dependencies.
+                _hasReachableVisibleUnresolvedById = FindReachableVisibleUnresolvedDependencies();
+            }
+            else
             {
-                System.Diagnostics.Debug.Assert(
-                    string.Equals(id, dependency.Id),
-                    "dependenciesWorld dictionary entry keys must match their value's ids.");
+                // There are no visible and unresolved dependencies in the snapshot
+                _hasReachableVisibleUnresolvedById = null;
+            }
+
+            return;
 
-                if (!dependency.Resolved && dependency.Visible)
+            static (ImmutableArray<IDependency> TopLevelDependencies, Dictionary<string, IDependency> topLevelDependencyByPath, bool hasVisibleUnresolvedDependency) Scan(ImmutableDictionary<string, IDependency> dependenciesWorld, ITargetFramework targetFramework)
```

Comment:
I'm curious whether there is some advantage I'm not thinking of in using a static local method here, when it returns a ValueTuple, which is a struct with fields, but the code-gen for a non-static local would also be a struct with fields.

Obviously there is a stylistic difference, and ensuring that there is no capturing, so not suggesting it should change, of course.

Reply:
Originally I had a version of this in play where it was capturing and mutating outer scope variables, and it was hard to reason about. So I made it static.